File: src/dataset/typescript/rational-numbers_exercism-wackerow/openAI/ChatGPT4_rational-numbers_v3.ts
export default class Rational {
  constructor(private a: number, private b: number) {}

  private static gcd(x: number, y: number): number {
    while (y !== 0) {
      [x, y] = [y, x % y];
    }
    return Math.abs(x);
  }

  public reduce(rational: Rational = new Rational(this.a, this.b)): Rational {
    const gcd = Rational.gcd(rational.a, rational.b);
    const sign = rational.b < 0 ? -1 : 1;
    return new Rational((rational.a / gcd) * sign, (rational.b / gcd) * sign);
  }

  public add(rational: Rational): Rational {
    const numerator = this.a * rational.b + rational.a * this.b;
    const denominator = this.b * rational.b;
    return this.reduce(new Rational(numerator, denominator));
  }

  public sub(rational: Rational): Rational {
    return this.add(new Rational(-rational.a, rational.b));
  }

  public mul(rational: Rational): Rational {
    return this.reduce(new Rational(this.a * rational.a, this.b * rational.b));
  }

  public div(rational: Rational): Rational {
    if (rational.a === 0) throw new Error("Cannot divide by zero");
    return this.reduce(new Rational(this.a * rational.b, this.b * rational.a));
  }

  public abs(): Rational {
    return this.reduce(new Rational(Math.abs(this.a), Math.abs(this.b)));
  }

  public exprational(power: number): Rational {
    if (power === 0) return new Rational(1, 1);
    const absPower = Math.abs(power);
    const numerator = power > 0 ? this.a ** absPower : this.b ** absPower;
    const denominator = power > 0 ? this.b ** absPower : this.a ** absPower;
    return this.reduce(new Rational(numerator, denominator));
  }

  public expreal(base: number): number {
    return Math.pow(base, this.a / this.b);
  }
}

File: src/dataset/python/rational-numbers_exercism-ThomasZumsteg/openAI/ChatGPT4_rational_numbers_v3.py
from math import gcd


class Rational:
    __slots__ = ('numer', 'denom')

    def __init__(self, numer, denom):
        if denom == 0:
            raise ValueError("Denominator cannot be zero.")
        common = gcd(numer, denom)
        if denom < 0:
            common = -common
        self.numer = numer // common
        self.denom = denom // common

    def __eq__(self, other):
        return (self.numer, self.denom) == (other.numer, other.denom)

    def __repr__(self):
        return f'{self.numer}/{self.denom}'

    def __add__(self, other):
        numer = self.numer * other.denom + other.numer * self.denom
        denom = self.denom * other.denom
        return Rational(numer, denom)

    def __sub__(self, other):
        numer = self.numer * other.denom - other.numer * self.denom
        denom = self.denom * other.denom
        return Rational(numer, denom)

    def __mul__(self, other):
        return Rational(self.numer * other.numer, self.denom * other.denom)

    def __truediv__(self, other):
        return Rational(self.numer * other.denom, self.denom * other.numer)

    def __abs__(self):
        return Rational(abs(self.numer), abs(self.denom))

    def __pow__(self, power):
        if power < 0:
            return Rational(self.denom ** -power, self.numer ** -power)
        return Rational(self.numer ** power, self.denom ** power)

    def __rpow__(self, base):
        return base ** (self.numer / self.denom)